spectral_cube/cube_utils.py

```diff
@@ -750,3 +751,105 @@ def bunit_converters(obj, unit, equivalencies=(), freq=None):
     else:
         # Slice along first axis to return a 1D array.
         return factors[0]
+
+def combine_headers(header1, header2):
+    '''
+    Given two Header objects, this function returns a fits Header of the optimal wcs. 
+
+    Parameters
+    ----------
+    header1 : astropy.io.fits.Header
+        A Header.
+    header2 : astropy.io.fits.Header
+        A Header.
+
+    Returns
+    -------
+    header : astropy.io.fits.Header
+        A header object of a field containing both initial headers. 
+        
+    '''
+    
+    from reproject.mosaicking import find_optimal_celestial_wcs
+    
+    head1 = header1.copy()
+    head1['NAXIS'] = 2
+    del head1['NAXIS3']
+    head2 = header2.copy()
+    head2['NAXIS'] = 2
+    del head2['NAXIS3']
+    
+    # Get wcs and shape of both headers
+    w1 = WCS(head1, naxis=2)
+    s1 = w1.array_shape
+    w2 = WCS(head2, naxis=2)
+    s2 = w2.array_shape
+    
+    # Get the optimal wcs and shape for both fields together
+    wcs_opt, shape_opt = find_optimal_celestial_wcs([(s1, w1), (s2, w2)], auto_rotate=False)
+    
+    # Make a new header using the optimal wcs and information from cubes
+    header = header1.copy()
+    header['NAXIS'] = 3
+    header['NAXIS1'] = shape_opt[1]
+    header['NAXIS2'] = shape_opt[0]
+    header['NAXIS3'] = header1['NAXIS3']
+    header.update(wcs_opt.to_header())
+    header['WCSAXES'] = 3
+    return header 
+
+def mosaic_cubes(cubes, spectral_block_size=100):
+    '''
+    This function reprojects cubes onto a common grid and combines them to a single field.  
+
+    Parameters
+    ----------
+    cubes : iterable
+        Iterable list of SpectralCube objects to reproject and add together. 
+    spectral_block_size : int
+        Block size so that reproject does not run out of memory. 
+
+    Outputs
+    -------
+    cube : SpectralCube
+        A spectral cube with the list of cubes mosaicked together.
+    '''
+    
+    cube1 = cubes[0]
+    cube1.allow_huge_operations = True
+    header = cube1.header
+    
+    # Create a header for a field containing all cubes
+    for cu in cubes[1:]: 
+        header = combine_headers(header, cu.header)
+    
+    # Prepare an array and mask for the final cube
+    shape_opt = (header['NAXIS3'],header['NAXIS2'],header['NAXIS1'])
+    final_array = np.zeros(shape_opt)
+    mask_opt = np.zeros(shape_opt[1:])
+    
+    for cube in cubes:
+        # Reproject cubes to the header
+        cube_repr = cube.reproject(header, block_size=[spectral_block_size, cube.shape[1], cube.shape[2]])
+        
+        # Create weighting mask
+        mask = (cube_repr[0:1].get_mask_array()[0])
+        mask_opt += mask.astype(float)
+                
+        # Go through each slice of the cube, add it to the final array
+        for ii in range(final_array.shape[0]):
+            slice1 = np.nan_to_num(cube_repr[ii])
+            slice2 = final_array[ii]
+            final_array[ii] = np.add(final_array[ii], slice1)
```

Comment:
This is fine, but could also be
```suggestion
            final_array[ii] += slice1
```

Reply:
I think my way will avoid creating a new copy in memory